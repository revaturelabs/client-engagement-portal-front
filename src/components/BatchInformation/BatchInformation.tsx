import React, { useState } from 'react';
import { Card, CardBody, CardFooter, CardHeader, Col, Row } from 'reactstrap';
import "../../scss/BatchInformation.scss"
import "react-multi-carousel/lib/styles.css";
// import { AssociateCardFactory } from '../AssociateCard/AssociateCardFactory';
import PlanInterventionModalv2 from '../PlanInterventionModal/PlanInterventionModalv2';
import { Batch, batchSkillToImage } from '../../types';
/**
 * This component displays data about one specific batch.
 */
export const BatchInformation : React.FC<{batch:Batch}> = ({ batch }) => {
    const [image] = useState<string>(batchSkillToImage[batch.skill]); // Sets the image of this component to match the specialization

    return(
        <Row className="justify-content-center">
            <Col md="1" lg="2" />
            <Col md="10" lg="8">
                <div id="batch-info-wrapper">
                    <Card className="batch-info-card">
                        {/* Batch name header */}
                        <CardHeader>
                            <h4>{batch.name}</h4>
                        </CardHeader>
                        <CardBody>
                            {/* Core tech. label and image */}
                            <h5><b className = 'readctl'>Core Technology Learned:</b></h5>
                            <Row>
                                <Col className="logo-container">
                                <span className="helper"></span>
                                    <img src={image} alt="Specialization logo" className='speclogo' />
                                </Col>
                            </Row>
                            {/* Specialization text underneeth image */}
                            <Row style={{textAlign: "center"}}>
                                <Col><p id="test-skill">{batch.skill}</p></Col>
                            </Row>
                            {/* Trainer and end date text */}
                            <p id="test-train"><b className ='readtrainer'>Trainer:</b> {batch.trainer}</p>
                            <p id="test-end-date"><b className = 'readted'>Training End Date:</b> {batch.endDate}</p>
                        </CardBody>
                        <CardFooter />
                    </Card>
                    {batch.associateAssignments?.length
                        ? <>
                            {/* Cards which provide detailed info. on the batch associates */}
                            <h1>Batch Engineers: ({batch.associateAssignments.length})</h1>
                            {/* <AssociateCardFactory {...batch}/> */}
                        </>
                        : <span id="test-noAss"/>
                    }
                </div>
            </Col>
            <Col md="1" lg="2" />
            <PlanInterventionModalv2 />
        </Row>
    );
}
