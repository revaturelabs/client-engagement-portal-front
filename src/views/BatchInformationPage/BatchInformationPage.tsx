import React, { useState } from "react";
import { connect } from "react-redux";
import { Link, RouteComponentProps, withRouter } from "react-router-dom";
import { Container, DropdownItem, Spinner } from "reactstrap";
import { IBasicBatchInfo } from "../../components/BatchCard/BatchCard";
import { BatchInformation } from "../../components/BatchInformation/BatchInformation";
import { NavBar } from "../../components/NavBar/NavBar";
import { axiosInstance } from "../../util/axiosConfig";

interface IBatchId {
  batchId: string;
}

interface IBatchDetailedInfo {
  name: string;
  endDate: string;
  employeeAssignments: [
    {
      employee: {
        firstName: string;
        lastName: string;
      };
    }
  ];
  skill: string;
  associateAssignments: {
    active: boolean;
    associate: {
      firstName: string;
      lastName: string;
      grades: {
        gradeId: number;
        dateReceived: string;
        score: number;
      }[];
    };
  }[];
}

interface IProps extends RouteComponentProps<IBatchId>, IBasicBatchInfo {
  batches: [IBatchDetailedInfo];
}

export const BatchInformationPage: React.FC<IProps> = (props: IProps) => {
  const [hasSpinner, setSpinner] = useState(false);
  const [hasData, setRecievedData] = useState(false);
  const [batchData, setBatchData] = useState({
    name: "N/A",
    endDate: "",
    employeeAssignments: [
      {
        employee: {
          firstName: "N/A",
          lastName: "",
        },
      },
    ],
    skill: "N/A",
    associateAssignments: [],
  } as IBatchDetailedInfo);

  /**
   * This function gets all of the batch data from our back end. This
   * includes data about each associate's test / quiz scores.
   *
   * @param batchId the batch id passed in from the batch card on the
   * home page
   *
   * @returns This function just changes the batch state
   */
  const getBatchData = async (batchId: string) => {
    setSpinner(true);

    //array to place batch data into
    let batchInfo: IBatchDetailedInfo = {
      name: "N/A",
      endDate: "",
      employeeAssignments: [
        {
          employee: {
            firstName: "N/A",
            lastName: "",
          },
        },
      ],
      skill: "N/A",
      associateAssignments: [],
    };

    //get data from server based on user id that was given
    try {
      const response = await (await axiosInstance()).get(
        "/client/batch/" + batchId
      );

      batchInfo = { ...response.data };
    } catch (error: any) {
      console.log(error);
    } finally {
      setSpinner(false);
    }

    return batchInfo;
  };

  const getBatchDataNow = async () => {
    if (props.match.params.batchId != null) {
      getBatchData(
        props.match.params.batchId
      ).then((response: IBatchDetailedInfo) => setBatchData(response));
    }
    setRecievedData(true);
  };

  if (!hasData) {
    getBatchDataNow();
  }

  return (
    <>
      <Container
        style={{
          minHeight: "100vh",
          maxWidth: "100vw",
          backgroundColor: "#E3E3E3",
        }}
      >
        <NavBar>
          <Link to="/home">
            <DropdownItem>Return to Client Home</DropdownItem>
          </Link>
        </NavBar>

        {/* Spinner displays below nav bar */}

        {hasSpinner ? (
          <div className=" row justify-content-center">
            <Spinner color="info" style={{ margin: 70 }} />
          </div>
        ) : (
          <BatchInformation
            batches={[
              {
                batchId: props.match.params.batchId,
                batchName: batchData.name,
                endDate: batchData.endDate,
                skill: batchData.skill,
                trainer: `${batchData.employeeAssignments[0].employee.firstName} ${batchData.employeeAssignments[0].employee.lastName}`,
                associateAssignments: batchData.associateAssignments,
              },
            ]}
          />
        )}
      </Container>
    </>
  );
};

const mapStateToProps = (store: any) => {
  return {
    batches: store.batchState.batches,
  };
};

export default withRouter(connect<any>(mapStateToProps)(BatchInformationPage));
