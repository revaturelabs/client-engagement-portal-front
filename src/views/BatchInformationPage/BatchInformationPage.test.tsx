import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { BatchInformationPage } from "./BatchInformationPage";
import configureStore from "redux-mock-store";
import Axios from "axios";
import { act as testUtilsAct } from "react-dom/test-utils";
// import { act as testRendererAct } from "react-test-renderer";
import { mount } from "enzyme";
import { BatchInformation } from "../../components/BatchInformation/BatchInformation";
// import { AssociateCardFactory } from "../../components/AssociateCard/AssociateCardFactory";
import { createBrowserHistory, History } from "history";
import { BrowserRouter, Router } from "react-router-dom";

const mockStore = configureStore([]);
let fakeData: any;
let initialData: any;
let nullData : any;
let store: any;

const mockAxiosGet = jest.spyOn(Axios, "get");

beforeAll(() => {
  const ax = require("../../util/axiosConfig");
  ax.axiosInstance = jest.fn(() => {
    return Axios;
  });

  mockAxiosGet.mockImplementation(() => {
    return Promise.resolve({
      data : [...fakeData]
    });
  });

  fakeData = {
    match: {
      params: {
        batchId: "1",
      },
    },
    batches: [
      {
        name: "Test",
        endDate: "endDate",
        employeeAssignments: [
          {
            employee: {
              firstName: "TestFName",
              lastName: "TestLameName",
            },
          },
        ],
        skill: "Java, React, Spring",
        associateAssignments: [
          {
            active: "true",
            associate: {
              firstName: "wack",
              lastName: "jobs",
              grades: [
                {
                  gradeId: "GR6953",
                  dateReceived: "10/03/2020",
                  score: "5",
                },
              ],
            },
          },
        ],
      },
    ],
  };

  initialData = {
    match: {
      params: {
        batchId: "1",
      },
    },
    batches: [
      {
        name: "",
        endDate: "",
        employeeAssignments: [
          {
            employee: {
              firstName: "",
              lastName: "",
            },
          },
        ],
        skill: "",
        associateAssignments: [
          {
            active: false,
            associate: {
              firstName: "",
              lastName: "",
              grades: [
                {
                  gradeId: "",
                  dateReceived: "",
                  score: "",
                },
              ],
            },
          },
        ],
      },
    ],
  };
  nullData = {
    match: {
      params: {
        batchId: "1",
      },
    },
    batches: [
      {
        name: "",
        endDate: "",
        employeeAssignments: null,
        skill: "",
        associateAssignments: null,
      },
    ],
  };
});

describe("BatchInformationPage view", () => {
  let component: any;

  beforeEach(() => {
    store = mockStore({
      userState: {
        user: {
          firstName: "wack",
          lastName: "jobs",
        },
      },
      batchState: {
          batches: nullData.batches
      }
    });

    component = renderer.create(
    <BrowserRouter>
      <Provider store={store}>
        <BatchInformationPage {...initialData} />
      </Provider>
      </BrowserRouter>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});

it("should test axios call", async () => {
    let wrapper2: any;
    await testUtilsAct(async () => {
      wrapper2 = mount(    <BrowserRouter>
        <Provider store={store}>
          <BatchInformationPage {...initialData} />
        </Provider>
        </BrowserRouter>
    )});   

    wrapper2.update();
    wrapper2.setProps(fakeData);
    expect(wrapper2.find(BatchInformation).children().length).toBeGreaterThan(0);
  });

  it("should set trainer to N/A and associateAssignments to an empty array" , async () => {
    let wrapper2: any;
    let history: History;
    history = createBrowserHistory();
    history.push("/");
    await testUtilsAct(async () => {
        wrapper2 = mount(
            <Router history={history}>
          <Provider store={store}>
            <BatchInformationPage {...nullData} match={{params: {batchId: "1"}}}/>
          </Provider>
          </Router>
          )
    });   
    wrapper2.render();
    wrapper2.setProps({});
    /* Doesn't make sense with application logic */
    // expect(wrapper2.find(AssociateCardFactory).props()).toStrictEqual({"associateAssignments": [], "batchId": "1", "name": "N/A", "endDate": "N/A", "skill": "N/A", "trainer": "N/A "});
    
    // expect(wrapper2.find(AssociateCardFactory).length).toBe(0);
  });