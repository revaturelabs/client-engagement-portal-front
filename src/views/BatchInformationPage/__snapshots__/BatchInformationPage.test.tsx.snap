// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchInformationPage view should render with given state from Redux store 1`] = `
<div
  className="container"
  style={
    Object {
      "backgroundColor": "#E3E3E3",
      "maxWidth": "100vw",
      "minHeight": "100vh",
    }
  }
>
  <div
    className="justify-content-around myNav row"
  >
    <div
      className="justify-content-start logoContainer col-auto"
    >
      <img
        alt="revature logo"
        className="myLogo"
        id="revLogo"
        src=""
      />
    </div>
    <div
      className="d-flex align-items-center justify-content-end auto test1 col"
    >
      <div
        className="btn-group"
        onKeyDown={[Function]}
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          aria-label={null}
          className="btn btn-secondary"
          id="navDropButton"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "white",
              "border": "none",
              "margin": "10px",
            }
          }
          type="button"
        >
          <span
            className="myDropdown"
            id="usersName"
            style={
              Object {
                "backgroundColor": "white",
                "border": "none",
                "color": "#474C55",
                "margin": "5px",
              }
            }
          >
            Welcome, 
            wack jobs
             â–¼
          </span>
          <span
            className="myMobileDropdown"
          >
            <div
              className="hamburger-react"
              onClick={[Function]}
              onKeyUp={[Function]}
              role="button"
              style={
                Object {
                  "cursor": "pointer",
                  "height": "48px",
                  "outline": "none",
                  "position": "relative",
                  "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
                  "userSelect": "none",
                  "width": "48px",
                }
              }
              tabIndex={0}
            >
              <div
                style={
                  Object {
                    "background": "#474C55",
                    "height": "3px",
                    "left": "8px",
                    "position": "absolute",
                    "top": "13px",
                    "transform": "none",
                    "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
                    "width": "32px",
                  }
                }
              />
              <div
                style={
                  Object {
                    "background": "#474C55",
                    "height": "3px",
                    "left": "8px",
                    "position": "absolute",
                    "top": "23px",
                    "transform": "none",
                    "transition": "0.2s cubic-bezier(0, 0, 0, 1)",
                    "width": "32px",
                  }
                }
              />
              <div
                style={
                  Object {
                    "background": "#474C55",
                    "height": "3px",
                    "left": "8px",
                    "position": "absolute",
                    "top": "33px",
                    "transform": "none",
                    "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
                    "width": "32px",
                  }
                }
              />
            </div>
          </span>
        </button>
        <div
          aria-hidden={true}
          className="dropdown-menu dropdown-menu-right"
          role="menu"
          tabIndex="-1"
        >
          <a
            href="/home"
            onClick={[Function]}
          >
             
            <button
              className="dropdown-item"
              onClick={[Function]}
              role="menuitem"
              tabIndex="0"
              type="button"
            >
              Return to Client Home
            </button>
             
          </a>
          <h6
            className="dropdown-header"
            id="test-head"
            onClick={[Function]}
            tabIndex="-1"
          >
            Account Options
          </h6>
          <a
            href="/login"
          >
            <button
              className="dropdown-item"
              onClick={[Function]}
              role="menuitem"
              tabIndex="0"
              type="button"
            >
              Logout
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="justify-content-center row"
  >
    <div
      className="col-md-1 col-lg-2"
    />
    <div
      className="col-md-10 col-lg-8"
    >
      <div
        id="batch-info-wrapper"
      >
        <div
          className="batch-info-card card"
        >
          <div
            className="card-header"
          >
            <h4>
              N/A
            </h4>
          </div>
          <div
            className="card-body"
          >
            <h5>
              <b
                className="readctl"
              >
                Core Technology Learned:
              </b>
            </h5>
            <div
              className="row"
            >
              <div
                className="logo-container col"
              >
                <span
                  className="helper"
                />
                <img
                  alt="Specialization logo"
                  className="speclogo"
                  src=""
                />
              </div>
            </div>
            <div
              className="row"
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <div
                className="col"
              >
                <p
                  id="test-skill"
                >
                  N/A
                </p>
              </div>
            </div>
            <p
              id="test-train"
            >
              <b
                className="readtrainer"
              >
                Trainer:
              </b>
               
              N/A 
            </p>
            <p
              id="test-end-date"
            >
              <b
                className="readted"
              >
                Training End Date:
              </b>
               
              N/A
            </p>
          </div>
          <div
            className="card-footer"
          />
        </div>
        <span
          id="test-noAss"
        />
      </div>
    </div>
    <div
      className="col-md-1 col-lg-2"
    />
    <button
      className="intervention-button"
      onClick={[Function]}
    >
      Request Intervention
    </button>
  </div>
</div>
`;
